### PR TITLE
Popover fixes on tablet

### PR DESCRIPTION
**Pull request by srdjanrist**
### Details
I have changed how popovers are shown on iPad medium and large devices. Since these are considered as native, sometimes popovers were shown bottom docked modals, instead of onscreen modals. Medium and large devices are considered as being "not-small" and this flag (of not being small) with a combination of native methods caused inconsistencies in different places that popovers are shown.

### Fixed Issues
$ https://github.com/Expensify/App/issues/7375

### Tests
_____
1. Click plus on list of all chats
2. Observe popover behavior
______
1. Open single chat
2. Click plus beside compose message
3. Observe popover behavior
____
1. Follow previous steps
2. Click attachment
3. Observe popover behavior
____
1. Open workspaces
2. Click three dots
3. Observe popover behavior
____
1. Open single workspace
2. Click on image
3. Observe popover behavior
____
1. Open user settings
2. Click on image to change it
3. Observe popover behavior
____
1. Open user settings
2. Click on payments
3. Click on "Add payment method"
4. Observe popover behavior
____

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I made sure to pull `main` before submitting my PR for review
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I clearly indicated the environment tests should be run in (Staging vs Production)
- [ ] I wrote testing steps that cover success & fail scenarios (if applicable)
- [ ] I ran the tests & they passed on **all platforms**
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified there are no console errors related to changes in this PR
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I added comments when the code was not self explanatory
    - [ ] I put all copy / text shown in the product in all `src/languages/*` files (if applicable)
    - [ ] I followed proper naming convention for platform-specific files (if applicable)
    - [x] I followed style guidelines (in [`Styling.md`](https://github.com/Expensify/App/blob/main/STYLING.md)) for all style edits I made
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)

#### PR Reviewer Checklist
- [ ] I verified the Author pulled `main` before submitting the PR
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the testing environment is mentioned in the test steps
- [ ] I verified testing steps cover success & fail scenarios (if applicable)
- [ ] I verified tests pass on **all platforms** & I tested again on all platforms
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified there are no console errors related to changes in this PR
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified comments were added when the code was not self explanatory
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files (if applicable)
    - [ ] I verified proper naming convention for platform-specific files was followed (if applicable)
    - [ ] I verified [style guidelines](https://github.com/Expensify/App/blob/main/STYLING.md) were followed
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)

### QA Steps
_____
1. Click plus on list of all chats
2. Verify that popover appears as modal on wide screens, and bottom docked on narrow screens
______
1. Open single chat
2. Click plus beside compose message
3. Verify that popover appears as modal on wide screens, and bottom docked on narrow screens
____
1. Follow previous steps
2. Click attachment
3. Verify that popover appears as modal on wide screens, and bottom docked on narrow screens
____
1. Open workspaces
2. Click three dots
3. Verify that popover appears as modal on wide screens, and bottom docked on narrow screens
____
1. Open single workspace
2. Click on image
3. Verify that popover appears as modal on wide screens, and bottom docked on narrow screens
____
1. Open user settings
2. Click on image to change it
3. Verify that popover appears as modal on wide screens, and bottom docked on narrow screens
____
1. Open user settings
2. Click on payments
3. Click on "Add payment method"
4. Verify that popover appears as modal on wide screens, and bottom docked on narrow screens
____

- [x] Verify that no errors appear in the JS console

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots

<details>
<summary>Web</summary>
1-web-plus
_
2-web-chat-plus
_
3-web-call-options
_
4-web-workspace-threedots
_
5-web-workspace-image
_
6-web-profile-image
_
7-web-payment
</details>

<details>
<summary>Mobile web</summary>
8-mobile-web-plus
_
9-mobile-web-chat-plus
_
10-mobile-web-call-options
_
11-mobile-web-workspace-threedots
_
12-mobile-web-workspace-image
_
13-mobile-web-profile-image
_
14-mobile-web-payment
_

</details>

<details>
<summary>Desktop</summary>

1-desktop-wide-plus
_
2-desktop-wide-chat-plus
_
3-desktop-wide-workspace-threedots
_
4-desktop-wide-workspace-image
_
5-desktop-wide-profile-image
_
6-desktop-wide-payment

</details>

<details>
<summary>iOS</summary>
1-iphone-plus
2-iphone-chat-plus
3-iphone-attachment
4-iphone-workspace-threedots
5-iphone-workspace-image
6-iphone-profile-image
7-iphone-payment
</details>

<details>
<summary>iPad Narrow</summary>
1-ipad-narrow-plus
2-ipad-narrow-chat-plus
3-ipad-narrow-attachment
4-ipad-narrow-workspace-threedots
5-ipad-narrow-workspace-image
6-ipad-narrow-profile-image
7-ipad-narrow-payment
</details>

<details>
<summary>iPad Wide</summary>
1-ipad-wide-plus
2-ipad-wide-chat-plus
3-ipad-wide-attachment
4-ipad-wide-workspace-threedots
5-ipad-wide-workspace-image
6-ipad-wide-profile-image
7-ipad-wide-payment
</details>

<details>
<summary>Android</summary>
1-android-plus
2-android-chat-plus
3-android-attachment
4-android-workspace-threedots
5-android-workspace-image
6-android-profile-image
7-android-payment
</details>

### CODE REVIEW

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Missing QA Steps. 

image

Fix it please.

**Comment by srdjanrist** [comment]
> Missing QA Steps.
> 
> image
> 
> Fix it please.

I have updated QA steps.

**Comment by parasharrajat** [comment]
image

This is still not fixed. Do not use MD link format. Directly use the link as `https://github.com/Expensify/App/issues/7375`

**Comment by srdjanrist** [comment]
> image
> 
> This is still not fixed. Do not use MD link format. Directly use the link as ` $ https://github.com/Expensify/App/issues/7375`

Updated.

**Comment by parasharrajat** [comment]
Thanks for the hard work on the PR @srdjanrist. Could you please help me understand the motive and your approach behind the changes you have done?

**Comment by srdjanrist** [comment]
> Thanks for the hard work on the PR @srdjanrist. Could you please help me understand the motive and your approach behind the changes you have done?

Looking into component `Popover` native implementation, you can see on line `20` and `21`, that depending on `fromSidebarMediumScreen` prop, either popover modal or bottom docked modal is used.
This issue occurs on iPads with diagonal greater than 9.7", and they are considered medium screen-sized.
Due to them being medium, they should behave like on web, meaning they should use popover modal, instead of bottom docked modal.
Adding property `fromSidebarMediumScreen` with proper value, I have resolved the issue, so that on the screens I mentioned, bottom-docked modal is not used, but a proper, popover modal is used in the problematic locations.

Simply adding property `fromSidebarMediumScreen` resolved the issue, since before
`props.fromSidebarMediumScreen ? CONST.MODAL.MODAL_TYPE.POPOVER : CONST.MODAL.MODAL_TYPE.BOTTOM_DOCKED` (this is taken from the code), it would fallback to bottom-docked modal, when this property wasn't set.

File: src/components/AttachmentPicker/index.native.js [hunk 1]
```diff
@@ -304,8 +304,9 @@ class AttachmentPicker extends Component {
                 <Popover
                     onClose={this.close}
                     isVisible={this.state.isVisible}
-                    anchorPosition={styles.createMenuPosition}
                     onModalHide={this.onModalHide}
+                    anchorPosition={this.props.anchorPosition}
+                    fromSidebarMediumScreen={!this.props.isSmallScreenWidth}
```
Comment (parasharrajat) [inline src/components/AttachmentPicker/index.native.js:309]: How is this prop related to this menu? This menu opens from the composer :heavy_plus_sign: .

File: src/styles/styles.js [hunk 2]
```diff
@@ -1052,6 +1052,10 @@ const styles = {
         paddingVertical: 12,
     },
 
+    createPaymentMethodMenuContainer: {
+        top: 380,
+    },
```
Comment (parasharrajat) [inline src/styles/styles.js:1057]: I don't see why do we need to pass the width now. If it was working fine on web without it.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
There are many lint errors. Please fix those. Are you following the contribution guidelines? Please run `npm run lint` before pushing chnages.

I think the simplest way to do it would be to modify the Popover Component so that we do
```js
 <Modal
            type={!props.isSidebarScreenWidth ? CONST.MODAL.MODAL_TYPE.POPOVER : CONST.MODAL.MODAL_TYPE.BOTTOM_DOCKED}
 ``` 
 instead of passing `fromSidebarMediumScreen` which is basically doing the same. 
 
 Also, this prop is not defined in the prop definitions. If there is a reason to keep it, please add prop-types for it in every component it is passed as a prop.

File: src/components/AddPaymentMethodMenu.js [hunk 3]
```diff
@@ -42,6 +42,7 @@ const AddPaymentMethodMenu = props => (
         isVisible={props.isVisible}
         onClose={props.onClose}
         anchorPosition={props.anchorPosition}
+        fromSidebarMediumScreen={!props.isSmallScreenWidth}
```
[the remark this answers is not in the document]
Comment (srdjanrist) [reply src/components/AddPaymentMethodMenu.js:45]: You are right. This will be removed.

File: src/styles/styles.js (hunk 2 above)
Comment (srdjanrist) [reply src/styles/styles.js:1057]: This is because `let position = getClickedElementLocation(nativeEvent);` has native implementation which returns just `{ bottom: 0, left: 0 }`

**Comment by srdjanrist** [comment]
> There are many lint errors. Please fix those. Are you following the contribution guidelines? Please run `npm run lint` before pushing chnages.
> 
> I think the simplest way to do it would be to modify the Popover Component so that we do
> 
> ```js
>  <Modal
>             type={!props.isSidebarScreenWidth ? CONST.MODAL.MODAL_TYPE.POPOVER : CONST.MODAL.MODAL_TYPE.BOTTOM_DOCKED}
> ```
> 
> instead of passing `fromSidebarMediumScreen` which is basically doing the same.
> 
> Also, this prop is not defined in the prop definitions. If there is a reason to keep it, please add prop-types for it in every component it is passed as a prop.

Yes. This could be done that way. Would you like this prop `isSidebarScreenWidth` to be passed from outer component, or should I wrap `Popover` with `withWindowDimensions`, so it has properties from window dimensions?

**Comment by parasharrajat** [comment]
Wrapping with `withWindowDimensions` is fine.

File: src/styles/styles.js (hunk 2 above)
Comment (parasharrajat) [reply src/styles/styles.js:1057]: Can't we use https://reactnative.dev/docs/0.64/pressevent to calculate the values?

Comment (srdjanrist) [reply src/styles/styles.js:1057]: @parasharrajat Thanks for the docs. 
Looking at the example:
```
{
    changedTouches: [PressEvent],
    identifier: 1,
    locationX: 8,
    locationY: 4.5,
    pageX: 24,
    pageY: 49.5,
    target: 1127,
    timestamp: 85131876.58868201,
    touches: []
}
```
we could theoretically calculate the top left coordinate of the pressed component, by using `pageX`, `pageY` and `locationX`, `locationY`. We would still need to add a height value of the pressed component to `y` coordinate, to move the popover to its bottom.
What do you think about that?

Comment (parasharrajat) [reply src/styles/styles.js:1057]: Ok. I think that's fine if height is needed let's calculate it.

File: src/components/AttachmentPicker/index.native.js (hunk 1 above)
Comment (srdjanrist) [reply src/components/AttachmentPicker/index.native.js]: The menu for attachment picker opens from the same position as the one from the composer ➕.

File: src/libs/getClickedElementLocation/index.native.js [hunk 4]
```diff
@@ -1,12 +1,13 @@
...
  *
+ * @param {Object} nativeEvent
  * @returns {Object}
  */
-function getClickedElementLocation() {
+function getClickedElementLocation(nativeEvent) {
     return {
-        bottom: 0,
-        left: 0,
+        bottom: nativeEvent.nativeEvent.pageY - nativeEvent.nativeEvent.locationY,
```
Comment (parasharrajat) [inline src/libs/getClickedElementLocation/index.native.js:9]: `nativeEvent.nativeEvent` looks confusing.

File: src/components/AvatarWithImagePicker.js [hunk 5]
```diff
@@ -186,7 +186,10 @@ class AvatarWithImagePicker extends React.Component {
                             : (
                                 <DefaultAvatar />
                             )}
-                        <AttachmentPicker type={CONST.ATTACHMENT_PICKER_TYPE.IMAGE}>
+                        <AttachmentPicker
+                            anchorPosition={styles.createMenuPositionProfile}
```
Comment (parasharrajat) [inline src/components/AvatarWithImagePicker.js:190]:
```suggestion
                            anchorPosition={this.props.anchorPosition}
```

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/components/Popover/index.native.js [hunk 6]
```diff
@@ -3,7 +3,7 @@ import React from 'react';
 import {propTypes as popoverPropTypes, defaultProps} from './popoverPropTypes';
 import CONST from '../../CONST';
 import Modal from '../Modal';
-import {windowDimensionsPropTypes} from '../withWindowDimensions';
+import withWindowDimensions, {windowDimensionsPropTypes} from '../withWindowDimensions';
 
 const propTypes = {
     ...(_.omit(popoverPropTypes, _.keys(windowDimensionsPropTypes))),
```
[the remark this answers is not in the document]
Comment (srdjanrist) [reply src/components/Popover/index.native.js]: Updated. I have left only that propTypes from popoverPropTypes without omitting windowDimensionsProps

File: src/libs/getClickedElementLocation/index.native.js (hunk 4 above)
Comment (srdjanrist) [reply src/libs/getClickedElementLocation/index.native.js]: Agree. Based on previous handling of this, I have changed the code that calls this method to pass the nested `nativeEvent` property instead of the whole object.

File: src/components/AvatarWithImagePicker.js (hunk 5 above)
Comment (srdjanrist) [reply src/components/AvatarWithImagePicker.js]: Updated

File: src/components/AttachmentPicker/index.native.js [hunk 7]
```diff
@@ -304,8 +304,8 @@ class AttachmentPicker extends Component {
                 <Popover
                     onClose={this.close}
                     isVisible={this.state.isVisible}
-                    anchorPosition={styles.createMenuPosition}
                     onModalHide={this.onModalHide}
+                    anchorPosition={this.props.anchorPosition}
```
[the remark this answers is not in the document]
Comment (srdjanrist) [reply src/components/AttachmentPicker/index.native.js:308]: I have updated the prop definiton.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Changes are good. But there are a few things that need to be done. 
1. Check where `getClickedElementLocation` is used and Test all of those dropdowns. 
2. On Native, popover menus have different animation than the web.

**Comment by parasharrajat** [comment]
Web 

https://user-images.githubusercontent.com/24370807/160438003-f108738c-6b4e-432c-93ce-8a410b9f24e9.mp4

Android

https://user-images.githubusercontent.com/24370807/160438017-dc0e41f5-0218-49da-8d71-364c85345410.mp4


Does not look same to me?

**Comment by srdjanrist** [comment]
@parasharrajat 
Thanks for the videos.
I might have misunderstood something then.
The fadein-fadeout animation is only on native platforms.

Second video shows Android, and the animation is present. Is it not supposed to be there?

**Comment by parasharrajat** [comment]
The motive is to keep the same animation on all platforms. When Menu opens from bottom on Mobile native, the animation is differnt. But when they behave like a web on tablets. They should have same animations. 

cc: @Expensify/design

**Comment by shawnborton** [comment]
I agree with that Rajat.

**Comment by srdjanrist** [comment]
Thanks for the input!
I have updated the pull request with a new commit that handles animations on smaller screen sizes on web.
My tests have went over few devices with various sizes, and I couldn't find any inconsistencies now.
Please, double check this @parasharrajat, so that we can proceed.

File: src/components/Modal/BaseModal.js [hunk 8]
```diff
@@ -110,8 +110,8 @@ class BaseModal extends PureComponent {
                 animationOut={this.props.animationOut || animationOut}
                 useNativeDriver={this.props.useNativeDriver}
                 hideModalContentWhileAnimating={this.props.hideModalContentWhileAnimating}
-                animationInTiming={this.props.animationInTiming}
-                animationOutTiming={this.props.animationOutTiming}
+                animationInTiming={this.props.disableAnimation ? 1 : this.props.animationInTiming}
+                animationOutTiming={this.props.disableAnimation ? 1 : this.props.animationOutTiming}
```
Comment (parasharrajat) [inline src/components/Modal/BaseModal.js:114]: Not happy with this change. Can't we just follow the same pattern as we do on the web?

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/components/Modal/BaseModal.js (hunk 8 above)
Comment (srdjanrist) [reply src/components/Modal/BaseModal.js]: Updated the pull request with a recommended change. 😊

**Review by parasharrajat — COMMENTED** [review]
The code looks fine. I will test it shortly.

**Comment by parasharrajat** [comment]
Testing it. Finally fixed my Mac.

**Comment by parasharrajat** [comment]
Didn't complete my testing. I will finish it up tomorrow.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Issue on Android Tablet. 
image

Steps: 
1. Open the payments page. 
2. Click at the middle of any payment method. 
3. Menu is off the screen.


Also pleas merge main.

**Comment by srdjanrist** [comment]
<details>
<summary>Comment</summary>
> Issue on Android Tablet. image
> 
> Steps:
> 
> 1. Open the payments page.
> 2. Click at the middle of any payment method.
> 3. Menu is off the screen.
> 
> Also pleas merge main.
</details>

______
I have looked into this. 
What I noticed is that depeneding on which part of the menu item is clicked, parameters from onPress callback are different. Actually, besides `locationX` and `locationY`, also `target` is different.

Touching the "Add payment method" part yields:
` "locationX": 4.11676025390625, 
"locationY": 9.80316162109375,
"target": 27237`
image

Touching anywhere beside the "Add payment method" part yields:
`"locationX": 351.5091552734375, 
"locationY": 11.81427001953125,
"target": 27247`
As you can see the `target` property is different as well.

By using this location and on-screen location, the popover's position is calculated.
Due to positions being different (depending on where the menu item is clicked), I have no way to precisely and consistently calculate the position of popover.

Have you had this situation before? Do you have any suggestions for improvement?

**Comment by Julesssss** [comment]
Hey @srdjanrist, we'll need to resolve the issue @parasharrajat mentioned, as it doesn't look ideal. I'm afraid I don't have any suggestions for you, but sharing the issue in the shared [Slack channel](https://expensify.slack.com/archives/C01GTK53T8Q) might be a good idea. As it's quite possible that someone has faced and solved this problem before.

**Comment by srdjanrist** [comment]
I have consulted people on Slack, and will try their approach as well.

I was thinking of another approach to this.
@parasharrajat  Would it be fine to update ```MenuItem``` component, refactor a little bit, and make a transparent View, positioned over the whole menu item. That way, this new overlay view would take all touch events and positioning would be always relative to the whole menu item, instead of each separate child view individually?

**Comment by parasharrajat** [comment]
That would be Overkill. Can't we use `measureInWindow` to get the coordinates?

**Comment by srdjanrist** [comment]
> That would be Overkill. Can't we use `measureInWindow` to get the coordinates?

I will check this today.

**Comment by srdjanrist** [comment]
> > That would be Overkill. Can't we use `measureInWindow` to get the coordinates?
> 
> I will check this today.

Updated MenuItem with code for measuring layout with the mentioned method

File: src/components/MenuItem.js [hunk 9]
```diff
@@ -41,43 +41,63 @@ const defaultProps = {
     interactive: true,
 };
 
-const MenuItem = props => (
-    <Pressable
-        onPress={(e) => {
```
Comment (parasharrajat) [inline src/components/MenuItem.js:46]: This change is not acceptable. We don't have to touch the MenuItem for this. Instead, use the ref to calculate whatever is needed on the parent components that use it.

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/components/MenuItem.js (hunk 9 above)
Comment (Julesssss) [reply src/components/MenuItem.js:46]: Hi @srdjanrist, could you please address this comment so we can move forward :)
